tests/qml/helper.cpp

```diff
@@ -0,0 +1,108 @@
+  // of my code.
+  return instance()->m_mainWindowLoadedCalled;
+}
+
+void TestHelper::setMainWindowLoadedCalled(bool val) {
+  m_mainWindowLoadedCalled = val;
+}
+
+void TestHelper::triggerInitializeGlean() {
+  emit MozillaVPN::instance()->initializeGlean();
+}
+
+void TestHelper::triggerRecordGleanEvent(const QString& event) {
+  emit MozillaVPN::instance()->recordGleanEvent(event);
+}
+
+void TestHelper::triggerSetGleanSourceTags(const QStringList& tags) {
+  emit MozillaVPN::instance()->setGleanSourceTags(tags);
+}
+
+void TestHelper::triggerAboutToQuit() {
+  emit MozillaVPN::instance()->aboutToQuit();
+}
+
+// The next two functions duplicate mozillavpn.h logic, but I'm not
+// sure how else to generate an "expected" value for tst_mainWindowGlean.qml
+// as it needs to be dynamic based on the platform people are running tests on.
+// At the least this should fail give someone pause for thought if they change
+// the mozillavpn.h logic.
+QString TestHelper::osVersion() { return QSysInfo::productVersion(); }
+QString TestHelper::architecture() {
+  return QSysInfo::currentCpuArchitecture();
+}
+
+void TestHelper::qmlEngineAvailable(QQmlEngine* engine) {
+  engine->addImportPath("qrc:///compat");
+  engine->addImportPath("qrc:///components");
+  engine->addImportPath("qrc:///glean");
+  engine->addImportPath("qrc:///themes");
+  engine->addImportPath("qrc:///");
+
+  qmlRegisterSingletonType<MozillaVPN>(
+      "Mozilla.VPN", 1, 0, "VPN", [](QQmlEngine*, QJSEngine*) -> QObject* {
+        QObject* obj = MozillaVPN::instance();
+        QQmlEngine::setObjectOwnership(obj, QQmlEngine::CppOwnership);
+        return obj;
+      });
+
+  qmlRegisterSingletonType<MozillaVPN>(
+      "Mozilla.VPN", 1, 0, "VPNl18n", [](QQmlEngine*, QJSEngine*) -> QObject* {
+        QObject* obj = L18nStrings::instance();
+        QQmlEngine::setObjectOwnership(obj, QQmlEngine::CppOwnership);
+        return obj;
+      });
+
+  qmlRegisterSingletonType<MozillaVPN>(
+      "Mozilla.VPN", 1, 0, "VPNSettings",
+      [](QQmlEngine*, QJSEngine*) -> QObject* {
+        QObject* obj = SettingsHolder::instance();
+        QQmlEngine::setObjectOwnership(obj, QQmlEngine::CppOwnership);
+        return obj;
+      });
+
+  qmlRegisterSingletonType<TestHelper>(
+      "TestHelper", 1, 0, "TestHelper",
+      [this](QQmlEngine*, QJSEngine*) -> QObject* {
+        QObject* obj = this;
+        QQmlEngine::setObjectOwnership(obj, QQmlEngine::CppOwnership);
+        return obj;
+      });
+
+  qmlRegisterSingletonType<MozillaVPN>(
+      "Mozilla.VPN", 1, 0, "VPNCloseEventHandler",
+      [this](QQmlEngine*, QJSEngine*) -> QObject* {
+        QObject* obj = &this->closeEventHandler;
+        QQmlEngine::setObjectOwnership(obj, QQmlEngine::CppOwnership);
+        return obj;
+      });
+
+  qmlRegisterSingletonType<MozillaVPN>(
+      "Mozilla.VPN", 1, 0, "VPNWhatsNewModel",
+      [this](QQmlEngine*, QJSEngine*) -> QObject* {
+        QObject* obj = this->whatsNewModel();
```

Comment:
Why here a method and for `closeEventHandler` a variable member? This can be: `QObject* obj = m_whatsNewModel;`

Reply:
along the way I got confused about how to create things in order only after initialization steps had been done. I now know and can make this consistent.